### PR TITLE
[#1] TPGM Benchmarks added

### PR DESCRIPTION
**Pull request by ChrizZz110**
Added 3 Benchmark classes:
* Aggregation
* Snapshot
* Difference

### CODE REVIEW

File: src/main/java/org/gradoop/benchmarks/tpgm/BaseTpgmBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * Get the path with a separator char at the end.
+   *
+   * @param path the path to append the separator if it do not have it
+   * @return the path as string with a separator at the end
+   */
+  private static String getPath(String path) {
+    return path.endsWith(File.separator) ? path : path + File.separator;
+  }
```
Comment (p-f) [inline src/main/java/org/gradoop/benchmarks/tpgm/BaseTpgmBenchmark.java]: Is this necessary? Wouldn't `path + File.separator` be sufficient? Multiple separators should still form a valid path (at least for POSIX-compatible systems)

File: src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  static {
+    OPTIONS.addRequiredOption(OPTION_QUERY_1, "queryts1", true, "Used first query timestamp [ms]");
+    OPTIONS.addRequiredOption(OPTION_QUERY_2, "queryts2", true, "Used second query timestamp [ms]");
+    OPTIONS.addOption(OPTION_VERIFICATION, "verification", false, "Verify Snapshot with join.");
+  }
+
+  /**
+   * Main program to run the benchmark. Arguments are the available options.
+   * Example: {@code /path/to/flink run -c org.gradoop.benchmark.tpgm.SnapshotBenchmark
+   * path/to/gradoop-examples.jar -i hdfs:///graph -o hdfs:///output -c results.csv
```
Comment (p-f) [inline src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java]: Again, I think this `gradoop-benchmarks` now

File: src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Reads the given arguments from command line
+   *
+   * @param cmd command line
+   */
+  private static void readCMDArguments(CommandLine cmd) {
+    QUERY_FROM_1 = Long.valueOf(cmd.getOptionValue(OPTION_QUERY_1));
+    QUERY_FROM_2 = Long.valueOf(cmd.getOptionValue(OPTION_QUERY_2));
```
Comment (p-f) [inline src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java]: I think you can store these parameters as `long` to implicitly avoid `null`, you can then use `Long.parseLong` instead.

File: src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        env.getLastJobExecutionResult().getNetRuntime(TimeUnit.SECONDS));
+
+    File f = new File(CSV_PATH);
+    if (f.exists() && !f.isDirectory()) {
+      FileUtils.writeStringToFile(f, tail, true);
+    } else {
+      PrintWriter writer = new PrintWriter(CSV_PATH, "UTF-8");
+      writer.print(head);
+      writer.print(tail);
+      writer.close();
```
Comment (p-f) [inline src/main/java/org/gradoop/benchmarks/tpgm/DiffBenchmark.java]: Same thing here, this looks odd with the 2 different ways of writing the file (also missing charset parameter of the `FileUtils` method)

File: src/main/java/org/gradoop/benchmarks/tpgm/SnapshotBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    OPTIONS.addRequiredOption(OPTION_QUERY_TYPE, "type", true, "Used query type");
+    OPTIONS.addOption(OPTION_VERIFICATION, "verification", false, "Verify Snapshot with join.");
+    OPTIONS.addOption(OPTION_QUERY_FROM, "from", true, "Used query from timestamp [ms]");
+    OPTIONS.addOption(OPTION_QUERY_TO, "to", true, "Used query to timestamp [ms]");
+  }
+
+  /**
+   * Main program to run the benchmark. Arguments are the available options.
+   * Example: {@code /path/to/flink run -c org.gradoop.benchmark.tpgm.SnapshotBenchmark
+   * path/to/gradoop-examples.jar -i hdfs:///graph -o hdfs:///output -c results.csv
```
Comment (p-f) [inline src/main/java/org/gradoop/benchmarks/tpgm/SnapshotBenchmark.java]: JAR name here also

File: src/main/java/org/gradoop/benchmarks/tpgm/AggregationBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    String tail = String
+      .format("%s|%s|%s%n",
+        env.getParallelism(),
+        INPUT_PATH,
+        env.getLastJobExecutionResult().getNetRuntime(TimeUnit.SECONDS));
+
+    File f = new File(CSV_PATH);
+    if (f.exists() && !f.isDirectory()) {
+      FileUtils.writeStringToFile(f, tail, true);
```
[the remark this answers is not in the document]
Comment (ChrizZz110) [reply src/main/java/org/gradoop/benchmarks/tpgm/AggregationBenchmark.java]: Good approach

File: src/main/java/org/gradoop/benchmarks/tpgm/BaseTpgmBenchmark.java (hunk 1 above)
Comment (ChrizZz110) [reply src/main/java/org/gradoop/benchmarks/tpgm/BaseTpgmBenchmark.java]: I will use the appendSeparator method from the AbstractRunner class.

File: pom.xml [hunk 7]
```diff
@@ -119,6 +119,12 @@
                         </goals>
                         <configuration>
                             <minimizeJar>true</minimizeJar>
+                            <relocations>
+                                <relocation>
+                                    <pattern>org.apache.commons.cli</pattern>
+                                    <shadedPattern>org.gradoop.examples.thirdparty.commons-cli</shadedPattern>
```
Comment (p-f) [inline pom.xml]: A `-` is not allowed as part of a Java identifier, see [Java Specification](https://docs.oracle.com/javase/specs/jls/se12/html/jls-3.html#jls-3.8)
While the JVM should not care about this, is could lead to problems when using related classes.
I think this should be `org.gradoop.examples.thirdparty.commons_cli` or `org.gradoop.examples.thirdparty.org.apache.commons.cli`

**Pull request merged**